src/plugins/data/common/search/expressions/esaggs/esaggs_fn.ts

```diff
@@ -0,0 +1,154 @@
+
+const name = 'esaggs';
+
+type Input = KibanaContext | null;
+type Output = Promise<Datatable>;
+
+interface Arguments {
+  index: string;
+  metricsAtAllLevels: boolean;
+  partialRows: boolean;
+  includeFormatHints: boolean;
+  aggConfigs: string;
+  timeFields?: string[];
+}
+
+export type EsaggsExpressionFunctionDefinition = ExpressionFunctionDefinition<
+  'esaggs',
+  Input,
+  Arguments,
+  Output
+>;
+
+/** @internal */
+export interface EsaggsStartDependencies {
+  addFilters?: AddFilters;
+  aggs: AggsStart;
+  deserializeFieldFormat: FormatFactory;
+  indexPatterns: IndexPatternsContract;
+  searchSource: ISearchStartSearchSource;
+}
+
+/** @internal */
+export const getEsaggsMeta: () => Omit<EsaggsExpressionFunctionDefinition, 'fn'> = () => ({
+  name,
+  type: 'datatable',
+  inputTypes: ['kibana_context', 'null'],
+  help: i18n.translate('data.functions.esaggs.help', {
+    defaultMessage: 'Run AggConfig aggregation',
+  }),
+  args: {
+    index: {
+      types: ['string'],
+      help: '',
+    },
+    metricsAtAllLevels: {
+      types: ['boolean'],
+      default: false,
+      help: '',
+    },
+    partialRows: {
+      types: ['boolean'],
+      default: false,
+      help: '',
+    },
+    includeFormatHints: {
+      types: ['boolean'],
+      default: false,
+      help: '',
+    },
+    aggConfigs: {
+      types: ['string'],
+      default: '""',
+      help: '',
+    },
+    timeFields: {
+      types: ['string'],
+      help: '',
+      multi: true,
+    },
+  },
+});
+
+/** @internal */
+export async function handleEsaggsRequest(
+  input: Input,
+  args: Arguments,
+  params: RequestHandlerParams
+): Promise<Datatable> {
+  const resolvedTimeRange = input?.timeRange && calculateBounds(input.timeRange);
+
+  const response = await handleRequest(params);
+
+  const table: Datatable = {
+    type: 'datatable',
+    rows: response.rows,
+    columns: response.columns.map((column) => {
+      const cleanedColumn: DatatableColumn = {
+        id: column.id,
+        name: column.name,
+        meta: {
+          type: column.aggConfig.params.field?.type || 'number',
+          field: column.aggConfig.params.field?.name,
+          index: params.indexPattern?.title,
+          params: column.aggConfig.toSerializedFieldFormat(),
+          source: name,
+          sourceParams: {
+            indexPatternId: params.indexPattern?.id,
+            appliedTimeRange:
+              column.aggConfig.params.field?.name &&
+              input?.timeRange &&
+              args.timeFields &&
+              args.timeFields.includes(column.aggConfig.params.field?.name)
+                ? {
+                    from: resolvedTimeRange?.min?.toISOString(),
+                    to: resolvedTimeRange?.max?.toISOString(),
+                  }
+                : undefined,
+            ...column.aggConfig.serialize(),
+          },
+        },
+      };
+      return cleanedColumn;
+    }),
+  };
+
+  return table;
+}
```

Comment:
I've split out a few things here that we can share on the client & server:
1. The meta for the expression function definition
1. The types for the function's dependencies & definition
1. The request handler & logic that converts a response to a datatable